src/coreclr/nativeaot/System.Private.CoreLib/src/System/InvokeUtils.cs

```diff
@@ -134,87 +134,341 @@ private static Exception ConvertOrWidenPrimitivesEnumsAndPointersIfPossible(obje
                 return CreateChangeTypeArgumentException(srcEEType, dstEEType);
             }
 
+            EETypeElementType srcElementType = srcEEType->ElementType;
+
+            void* rawDstValue = null;
+
+            Unsafe.SkipInit(out char charValue);
+            Unsafe.SkipInit(out sbyte sbyteValue);
+            Unsafe.SkipInit(out byte byteValue);
+            Unsafe.SkipInit(out short shortValue);
+            Unsafe.SkipInit(out ushort ushortValue);
+            Unsafe.SkipInit(out int intValue);
+            Unsafe.SkipInit(out uint uintValue);
+            Unsafe.SkipInit(out long longValue);
+            Unsafe.SkipInit(out ulong ulongValue);
+            Unsafe.SkipInit(out float floatValue);
+            Unsafe.SkipInit(out double doubleValue);
+
+            Unsafe.SkipInit(out dstObject);
+
+            // This switch supports the conversions present in the table below, in 'CanPrimitiveWiden'
             switch (dstElementType)
             {
                 case EETypeElementType.Boolean:
-                    dstObject = Convert.ToBoolean(srcObject);
+
+                    // Can only be bool here
+                    dstObject = srcObject;
                     break;
 
                 case EETypeElementType.Char:
-                    char charValue = Convert.ToChar(srcObject);
-                    dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, charValue) : charValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.Char:
+                            charValue = RuntimeHelpers.FastUnbox<char>(srcObject);
+                            break;
+                        case EETypeElementType.Byte:
+                            charValue = (char)RuntimeHelpers.FastUnbox<byte>(srcObject);
+                            break;
+                        case EETypeElementType.UInt16:
+                            charValue = (char)RuntimeHelpers.FastUnbox<ushort>(srcObject);
+                            break;
+                        default:
+                            goto Failure;
+                    }
+                    rawDstValue = &charValue;
                     break;
 
                 case EETypeElementType.SByte:
-                    sbyte sbyteValue = Convert.ToSByte(srcObject);
-                    dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, sbyteValue) : sbyteValue;
-                    break;
 
-                case EETypeElementType.Int16:
-                    short shortValue = Convert.ToInt16(srcObject);
-                    dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, shortValue) : shortValue;
+                    // Can only be sbyte here
```

Comment:
Is this correct? Can this be reboxing enum to primitive or vice versa? I think the exact matches are handled earlier.

Reply:
What do you mean? My understanding was that regardless of whether either value was an enum here, we're just copying the raw primitive value to that local and then `RuntimeHelpers.Box` will take care of both scenarios at the end, either just boxing the primitive or boxing it as the target enum type. Is that not correct? 🤔

Reply:
This comment was meant to be on the `bool` case. The bool-backed enums are partially supported. I am not sure whether we can get here with bool backed enum, but the current code handles the bool backed enums so it would keep it that way.

Reply:
Oh interesting. I thought we had removed bool-backed enum support in .NET 8. Anyway, fixed in cdccf695f38b5973ce7b3c40d4cd2d915e617915. It also allows us to remove that check on `rawDstValue` not being `null`, so that's also nice 😄